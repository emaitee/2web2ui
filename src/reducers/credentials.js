const initialState = {
  grants: [],
  keys: [],
  loadingGrants: false,
  loadingKeys: false,
  error: null
};

export default (state = initialState, { payload, type }) => {
  switch (type) {
    // LIST_API_KEYS
    case 'LIST_API_KEYS_PENDING': {
      return { ...state, loadingKeys: true, error: null };
    }

    case 'LIST_API_KEYS_SUCCESS': {
      return { ...state, loadingKeys: false, keys: payload };
    }

    case 'LIST_API_KEYS_FAIL': {
      return { ...state, loadingKeys: false, error: payload };
    }

    // LIST_GRANTS
    case 'LIST_GRANTS_PENDING': {
      return { ...state, loadingGrants: true, error: null };
    }

    case 'LIST_GRANTS_SUCCESS': {
      return { ...state, loadingGrants: false, grants: payload };
    }

    case 'LIST_GRANTS_FAIL': {
      return { ...state, loadingGrants: false, error: payload };
    }

    default: {
      return state;
    }
  }
};
