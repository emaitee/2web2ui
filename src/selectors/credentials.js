import { find, fromPairs, isEmpty, keyBy, map, size } from 'lodash';
import { createSelector } from 'reselect';

/*
 * generic credentials selectors
 */
const getApiKeys = (state) => state.credentials.keys;
const getGrantsArray = (state) => state.credentials.grants;
const getSubaccounts = (state) => state.subaccounts.list;
const getApiKeyId = (state, props) => props.match.params.id;

const getLoadingKeys = (state) => state.credentials.loadingKeys;
const getLoadingGrants = (state) => state.credentials.loadingGrants;
const getLoadingSubaccounts = (state) => state.subaccounts.listLoading;

export const getApiKey = createSelector(
  [getApiKeys, getApiKeyId],
  (apiKeys, id) => find(apiKeys, { id })
);

// Convert grants array to an object keyed by `grant.key`
export const getGrants = createSelector(getGrantsArray, (grants) =>
  keyBy(grants, 'key')
);

export const getLoading = createSelector(
  [getLoadingKeys, getLoadingGrants, getLoadingSubaccounts],
  (loadingKeys, loadingGrants, loadingSubaccounts) =>
    loadingKeys || loadingGrants || loadingSubaccounts
);

/*
 * ApiKeyForm selectors
 */
const getFormApiKey = (state, props) => props.apiKey || {};

export const getIsNew = createSelector(getFormApiKey, (apiKey) =>
  isEmpty(apiKey)
);

export const getInitialGrantsRadio = createSelector(
  [getGrants, getFormApiKey, getIsNew],
  (grants, apiKey, isNew) =>
    isNew || size(grants) <= size(apiKey.grants) ? 'all' : 'select'
);

export const getInitialSubaccount = createSelector(
  [getSubaccounts, getFormApiKey],
  (subaccounts, apiKey) => find(subaccounts, { id: apiKey.subaccount_id })
);

export const getInitialValues = createSelector(getFormApiKey, (apiKey) => {
  // lodash.map lets us treat undefined nicely.
  const grantsPairs = map(apiKey.grants, (grant) => [grant, 'true']);

  return {
    ...apiKey,
    grants: fromPairs(grantsPairs)
  };
});
