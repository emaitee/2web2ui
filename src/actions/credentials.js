import { formatKeyForRequest } from './helpers/credentials';
import sparkpostApiRequest from './helpers/sparkpostApiRequest';

export function listApiKeys() {
  return (dispatch, getState) => {
    if (getState().credentials.keys.length) {
      return;
    }

    return dispatch(
      sparkpostApiRequest({
        type: 'LIST_API_KEYS',
        meta: {
          method: 'GET',
          url: '/api-keys'
        }
      })
    );
  };
}

export function createApiKey(key) {
  return (dispatch, getState) =>
    dispatch({
      type: 'CREATE_API_KEY',
      meta: {
        method: 'POST',
        url: '/api-keys',
        ...formatKeyForRequest(key, getState)
      }
    });
}

export function deleteApiKey(id) {
  return {
    type: 'DELETE_API_KEY',
    meta: {
      method: 'DELETE',
      url: `/api-keys/${id}`
    }
  };
}

export function updateApiKey(id, key) {
  return (dispatch, getState) =>
    dispatch({
      type: 'UPDATE_API_KEY',
      meta: {
        method: 'PUT',
        url: `/api-keys/${id}`,
        ...formatKeyForRequest(key, getState)
      }
    });
}

export function listGrants() {
  return (dispatch, getState) => {
    if (getState().credentials.grants.length) {
      return;
    }

    dispatch(
      sparkpostApiRequest({
        type: 'LIST_GRANTS',
        meta: {
          method: 'GET',
          url: '/authenticate/grants'
        }
      })
    );
  };
}
