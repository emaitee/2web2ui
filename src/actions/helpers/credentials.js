import { pickBy, trim, keys } from 'lodash';

import { getGrants } from 'selectors/credentials';

export function formatKeyForRequest(key, getState) {
  const request = { data: {}};

  if (key.subaccount) {
    request.headers = { 'X-MSYS-SUBACCOUNT': key.subaccount.id };
  }

  request.data.label = key.label;

  if (key.grantsRadio === 'all') {
    request.data.grants = keys(getGrants(getState()));
  } else {
    request.data.grants = keys(pickBy(key.grants));
  }

  if (key.validIps) {
    request.data.valid_ips = key.validIps.split(',').map(trim);
  }

  return request;
}
