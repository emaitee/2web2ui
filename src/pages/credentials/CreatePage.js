import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Page, Panel } from '@sparkpost/matchbox';

import { createApiKey } from 'actions/credentials';
import Layout from 'components/layout/Layout';
import { getLoading } from 'selectors/credentials';
import ApiKeyForm from './components/ApiKeyForm';

const breadcrumbAction = {
  content: 'API Keys',
  Component: Link,
  to: '/account/credentials'
};

const CreatePage = ({ createApiKey, loading }) => (
  <Layout.App loading={loading}>
    <Page title="Create API Key" breadcrumbAction={breadcrumbAction} />
    <Panel>
      <Panel.Section>
        <ApiKeyForm onSubmit={createApiKey} />
      </Panel.Section>
    </Panel>
  </Layout.App>
);

const mapStateToProps = (state, props) => ({
  loading: getLoading(state)
});

export default connect(mapStateToProps, { createApiKey })(CreatePage);
