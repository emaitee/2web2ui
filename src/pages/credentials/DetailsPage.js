import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { Page, Panel } from '@sparkpost/matchbox';

import { deleteApiKey, listApiKeys, updateApiKey } from 'actions/credentials';
import Layout from 'components/layout/Layout';
import { getApiKey, getLoading } from 'selectors/credentials';
import ApiKeyForm from './components/ApiKeyForm';

const breadcrumbAction = {
  content: 'API Keys',
  Component: Link,
  to: '/account/credentials'
};

export class CredentialsDetailsPage extends Component {
  static defaultProps = {
    apiKey: {}
  };

  constructor(props) {
    super(props);

    this.secondaryActions = [
      { content: 'Delete', onClick: this.props.deleteApiKey }
    ];
  }

  componentDidMount() {
    this.props.listApiKeys();
  }

  render() {
    const { apiKey, loading } = this.props;

    return (
      <Layout.App loading={loading}>
        <Page
          title={apiKey.label}
          breadcrumbAction={breadcrumbAction}
          secondaryActions={this.secondaryActions}
        />
        <Panel>
          <Panel.Section>
            <ApiKeyForm apiKey={apiKey} onSubmit={this.props.updateApiKey} />
          </Panel.Section>
        </Panel>
      </Layout.App>
    );
  }
}

const mapStateToProps = (state, props) => {
  const { error, grants, keys } = state.credentials;

  return {
    apiKey: getApiKey(state, props),
    keys,
    error,
    grants,
    loading: getLoading(state)
  };
};

const mapDispatchToProps = (dispatch, props) => {
  const { id } = props.match.params;

  return {
    deleteApiKey: () => dispatch(deleteApiKey(id)),
    listApiKeys: () => dispatch(listApiKeys()),
    updateApiKey: (values) => dispatch(updateApiKey(id, values))
  };
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(CredentialsDetailsPage)
);
