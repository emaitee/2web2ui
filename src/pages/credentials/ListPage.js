import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Page } from '@sparkpost/matchbox';

import { listApiKeys } from 'actions/credentials';

import ApiErrorBanner from 'components/apiErrorBanner/ApiErrorBanner';
import TableCollection from 'components/collection/TableCollection';
import Layout from 'components/layout/Layout';
import { getLoading } from 'selectors/credentials';
import PermissionsColumn from './components/PermissionsColumn';

const columns = ['Name', 'Key', 'Permissions'];

const primaryAction = {
  content: 'Create API Key',
  Component: Link,
  to: '/account/credentials/create'
};

const getRowData = (key) => [
  <Link to={`/account/credentials/details/${key.id}`}>{key.label}</Link>,
  <code>{key.short_key} ••••</code>,
  <PermissionsColumn grants={key.grants} />
];

export class ListPage extends Component {
  componentDidMount() {
    this.props.listApiKeys();
  }

  renderCollection() {
    const { keys } = this.props;

    return (
      <TableCollection
        columns={columns}
        getRowData={getRowData}
        pagination={true}
        rows={keys}
      />
    );
  }

  renderError() {
    const { error, listApiKeys } = this.props;

    return (
      <ApiErrorBanner
        errorDetails={error.message}
        message="Sorry, we seem to have had some trouble loading your API keys."
        reload={listApiKeys}
      />
    );
  }

  render() {
    const { error, loading } = this.props;

    return (
      <Layout.App loading={loading}>
        <Page primaryAction={primaryAction} title="API Keys" />
        {error ? this.renderError() : this.renderCollection()}
      </Layout.App>
    );
  }
}

const mapStateToProps = (state) => {
  const { error, keys } = state.credentials;
  return {
    error,
    keys,
    loading: getLoading(state)
  };
};

export default connect(mapStateToProps, { listApiKeys })(ListPage);
