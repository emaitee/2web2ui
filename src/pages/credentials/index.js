import CreatePage from './CreatePage';
import DetailsPage from './DetailsPage';
import ListPage from './ListPage';

export default {
  CreatePage,
  DetailsPage,
  ListPage
};
