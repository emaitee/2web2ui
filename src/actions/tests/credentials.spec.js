import { createMockStore } from '../../config/tests/mockStore';
import * as Actions from '../credentials';

jest.mock('../helpers/sparkpostApiRequest', () => jest.fn((action) => action));

describe('listApiKeys()', () => {
  it('dispatches the correct action when no keys are present', () => {
    const store = createMockStore({ credentials: { keys: []}});

    store.dispatch(Actions.listApiKeys());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('does not dispatch when keys are present', () => {
    const store = createMockStore({
      credentials: { keys: [{ label: 'a' }]}
    });

    store.dispatch(Actions.listApiKeys());
    expect(store.getActions()).toHaveLength(0);
  });
});

describe('listGrants()', () => {
  it('dispatches the correct action when no grants are present', () => {
    const store = createMockStore({ credentials: { grants: []}});

    store.dispatch(Actions.listGrants());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('does not dispatch when grants are present', () => {
    const store = createMockStore({
      credentials: { grants: ['metrics/view']}
    });

    store.dispatch(Actions.listGrants());
    expect(store.getActions()).toHaveLength(0);
  });
});
