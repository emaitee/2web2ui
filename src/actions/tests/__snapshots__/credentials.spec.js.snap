// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listApiKeys() dispatches the correct action when no keys are present 1`] = `
Array [
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/api-keys",
    },
    "type": "LIST_API_KEYS",
  },
]
`;

exports[`listGrants() dispatches the correct action when no grants are present 1`] = `
Array [
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/authenticate/grants",
    },
    "type": "LIST_GRANTS",
  },
]
`;
